library/std/src/panicking.rs

```diff
@@ -27,6 +27,22 @@ use crate::sys::backtrace;
 use crate::sys::stdio::panic_output;
 use crate::{fmt, intrinsics, process, thread};
 
+// This forces codegen of the function called by panic!() inside the std crate, rather than in
+// downstream crates. Primarily this is useful for rustc's codegen tests, which rely on noticing
+// complete removal of panic from generated IR. Since begin_panic is inline(never), it's only
+// codegen'd once per crate-graph so this pushes that to std rather than our codegen test crates.
+//
+// (See https://github.com/rust-lang/rust/pull/123244 for more info on why).
+//
+// If this is causing problems we can also modify those codegen tests to use a crate type like
+// cdylib which doesn't export "Rust" symbols to downstream linkage units.
+#[unstable(feature = "libstd_sys_internals", reason = "used by the panic! macro", issue = "none")]
+#[doc(hidden)]
+#[allow(dead_code)]
+#[used]
```

Comment:
Maybe `#[used(compiler)]`? This shouldn't be `#[used(linker)]` as this shouldn't prevent `begin_panic` from getting optimized away if it is genuinely unused, like with `panic_immediate_abort`. `#[used]` is currently the same as `#[used(compiler)]`, but we may want to change it to `#[used(linker)]` as the latter is sometimes necessary to actually preserve the symbol after linking rather than just preventing LTO from discarding it.

Reply:
Hm, maybe. What we actually care about here is *not* this static, but rather the underlying begin_panic function getting codegen'd in libstd.so/rlib so it doesn't need to get codegen'd in downstream crates.

I *think* used(compiler) is sufficient for that.

Reply:
Indeed, in this case used(compiler) is enough. I'm proposing to use that explicitly in case `#[used]` gets changed to used(linker) in the future to avoid unnecessarily keeping the function around if that happens.

Reply:
Yep, I think latest changes made that adjustment already. Thanks for bringing it up!